File: Iterator/java/Iterator.java
class DataStack{
    private int[] items = new int[10];
    private int length;

    public DataStack(){
        length = -1;
    }

    public DataStack(DataStack myStack){
        this.items = myStack.items;
        this.length = myStack.length;
    }

    public int[] getItems(){
        return items;
    }

    public int getLength(){
        return length;
    }

    public void push(int value){
        items[++length] = value;
    }

    public int pop(){
        return items[length--];
    }

    public static boolean equals(DataStack myStack1, DataStack myStack2){
        StackIterator it1 = new StackIterator(myStack1),
                it2 = new StackIterator(myStack2);

        while (it1.isEnd() || it2.isEnd()){
            if (it1.get() != it2.get()) break;
            StackIterator.increment(it1);
            StackIterator.increment(it2);
        }

        return !it1.isEnd() && !it2.isEnd();
    }
}

class StackIterator
{
    private DataStack stack;
    private int index;

    public StackIterator(DataStack myStack){
        this.stack = myStack;
        this.index = 0;
    }

    public static StackIterator increment(StackIterator s){
        s.index++;
        return s;
    }

    public int get(){
        if (index < stack.getLength())
            return stack.getItems()[index];
        return 0;
    }

    public boolean isEnd(){
        return index != stack.getLength() + 1;
    }
}

public class Main {
    public static void main(String[] args){

        DataStack myStack1 = new DataStack();
        
        for (int i = 0; i < 5; i++){
            myStack1.push(i);
        }

        DataStack myStack2 = new DataStack(myStack1);

        System.out.println(DataStack.equals(myStack1, myStack2));

        myStack2.push(10);

        System.out.println(DataStack.equals(myStack1, myStack2));
    }
}

File: Iterator/python/Iterator.py
import copy


class DataStack:

    def __init__(self, my_stack: 'DataStack' = None):
        self.__items = [0] * 10
        self.__length = 0

        if my_stack is not None:
            self.__items = copy.deepcopy(my_stack.__items)
            self.__length = my_stack.__length

    @property
    def items(self):
        return self.__items

    @property
    def length(self) -> int:
        return self.__length

    def push(self, value: int):
        self.__items[self.__length] = value
        self.__length += 1

    def pop(self) -> int:
        self.__length -= 1
        return self.__items[self.__length]

    def __eq__(self, other: 'DataStack') -> bool:
        it1, it2 = StackIterator(self), StackIterator(other)

        while not it1.is_end() or not it2.is_end():
            if next(it1) != next(it2):
                break
        return it1.is_end() and it2.is_end()


class StackIterator:

    def __init__(self, my_stack: DataStack):
        self.__stack = my_stack
        self.__index = 0

    def __iter__(self):
        return self

    def __next__(self):
        current_index = self.__index
        self.__index += 1
        if current_index < self.__stack.length:
            return self.__stack.items[current_index]
        return 0

    def is_end(self) -> bool:
        return self.__index == self.__stack.length + 1


if __name__ == '__main__':

    my_stack1 = DataStack()
    for i in range(1, 5):
        my_stack1.push(i)

    my_stack2 = DataStack(my_stack1)

    print(my_stack1 == my_stack2)

    my_stack2.push(10)

    print(my_stack1 == my_stack2)